File: LLD/SplitwiseSystem.java
package com.splitwise.lld;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/*
 * ==============================================================================================
 * SPLITWISE LOW LEVEL DESIGN
 * ==============================================================================================
 * 
 * Key Features:
 * 1. User/Group Management: Create users, create groups, add members.
 * 2. Expense Management: Add expenses with Equal, Exact, or Percentage splits.
 * 3. Balance Calculation: Automatically calculate "Who owes Whom" (Positive/Negative offsets).
 * 4. Settlement: Logic for users to pay back and clear debts.
 * 
 * Design Patterns:
 * 1. Singleton: SplitwiseService (Facade).
 * 2. Strategy: SplitStrategy (HANDLES Equal, Exact, Percent logic).
 * 3. Factory: SplitFactory (Creates Split objects).
 *
 * Class Design Diagram:
 * ---------------------
 * [SplitwiseService] "1" *-- "*" [User]
 * [SplitwiseService] "1" *-- "*" [Group]
 * [Group] "1" *-- "*" [Expense]
 * [Expense] "1" *-- "*" [Split]
 * [Expense] "1" *-- "1" [User] (PaidBy)
 * [Split] <|-- [EqualSplit]
 * [Split] <|-- [ExactSplit]
 * [Split] <|-- [PercentSplit]
 * [User] ..> [BalanceSheet] (Map<User, Double>)
 *
 * Class Details:
 * ---------------------
 * 1. SplitwiseService (Singleton)
 *    - Role: Main controller.
 *    - Methods: addExpense(), settleBalance(), showBalance().
 * 
 * 2. Expense
 *    - Role: Represents a transaction.
 *    - Attributes: amount, paidBy, splits (List), splitType.
 *
 * 3. Split (Abstract)
 *    - Role: Represents a share of an expense.
 *    - Attributes: user, amount.
 *    - Subclasses: EqualSplit, ExactSplit, PercentSplit (has 'percent' field).
 *
 * 4. SplitType (Enum)
 *    - Types: EQUAL, EXACT, PERCENT.
 *
 * 5. User
 *    - Role: Participant.
 *    - Attributes: id, name, email.
 */

public class SplitwiseSystem {
    public static void main(String[] args) {
        System.out.println("--- Splitwise System Demo ---");
        
        SplitwiseService service = SplitwiseService.getInstance();

        // 1. Create Users
        User u1 = new User("U1", "Alice", "alice@test.com", "9999");
        User u2 = new User("U2", "Bob", "bob@test.com", "8888");
        User u3 = new User("U3", "Charlie", "charlie@test.com", "7777");
        
        service.addUser(u1); service.addUser(u2); service.addUser(u3);

        // 2. Create Group
        Group g1 = new Group("G1", "Trip");
        g1.addMember(u1); g1.addMember(u2); g1.addMember(u3);
        service.addGroup(g1);

        // 3. Add Expense: EQUAL SPLIT
        // Alice pays 300, split equally among Alice, Bob, Charlie (100 each)
        System.out.println("\n[Transaction] Alice pays 300 for Lunch (Equal Split)");
        List<Split> splits1 = new ArrayList<>();
        splits1.add(new EqualSplit(u1));
        splits1.add(new EqualSplit(u2));
        splits1.add(new EqualSplit(u3));
        
        service.addExpense("Lunch", 300, u1, splits1, SplitType.EQUAL);
        
        service.showBalance(u2.getId()); // Bob should owe Alice 100
        service.showBalance(u3.getId()); // Charlie should owe Alice 100

        // 4. Add Expense: EXACT SPLIT
        // Bob pays 1000, Alice owes 300, Charlie owes 700. (Bob owes 0 to self)
        System.out.println("\n[Transaction] Bob pays 1000 for Cab (Exact: Alice 300, Charlie 700)");
        List<Split> splits2 = new ArrayList<>();
        splits2.add(new ExactSplit(u1, 300));
        splits2.add(new ExactSplit(u3, 700));
        
        service.addExpense("Cab", 1000, u2, splits2, SplitType.EXACT);
        
        // 5. Add Expense: PERCENT SPLIT
        // Charlie pays 500. Alice 50%, Bob 20%, Charlie 30%.
        System.out.println("\n[Transaction] Charlie pays 500 for Party (Percent: Alice 50%, Bob 20%, Charlie 30%)");
        List<Split> splits3 = new ArrayList<>();
        splits3.add(new PercentSplit(u1, 50.0));
        splits3.add(new PercentSplit(u2, 20.0));
        splits3.add(new PercentSplit(u3, 30.0));
        
        service.addExpense("Party", 500, u3, splits3, SplitType.PERCENT);

        // 6. Show Final Balances
        System.out.println("\n--- Balances Before Settlement ---");
        service.showBalance(u1.getId());
        
        // 7. Settle Debt
        // Alice owes Charlie some amount. Let's say she pays 200 to Charlie.
        System.out.println("\n[Transaction] Alice pays 200 to Charlie to settle part of the debt");
        service.settleBalance(u1, u3, 200.0);
        
        System.out.println("\n--- Final Balances After Settlement ---");
        service.showBalance(u1.getId());
        service.showBalance(u2.getId());
        service.showBalance(u3.getId());
    }
}

// ==========================================
// Enums & Strategy
// ==========================================

enum SplitType { EQUAL, EXACT, PERCENT }

abstract class Split {
    User user;
    double amount;
    
    public Split(User user) { this.user = user; }
    public double getAmount() { return amount; }
    public void setAmount(double amount) { this.amount = amount; }
    public User getUser() { return user; }
}

class EqualSplit extends Split {
    public EqualSplit(User user) { super(user); }
}

class ExactSplit extends Split {
    public ExactSplit(User user, double amount) { 
        super(user); 
        this.amount = amount;
    }
}

class PercentSplit extends Split {
    double percent;
    public PercentSplit(User user, double percent) { 
        super(user); 
        this.percent = percent;
    }
    public double getPercent() { return percent; }
}

// ==========================================
// Domain Models
// ==========================================

class User {
    private String id;
    private String name;
    
    // Helper to print nice output (Who owes Whom)
    // Map<OtherUserID, Amount> -> +ive means User gets back, -ive means User owes
    // Actually, simpler is: userExpenseSheet: Map<User, Double>
    
    public User(String id, String name, String email, String phone) {
        this.id = id; this.name = name;
    }
    public String getId() { return id; }
    public String getName() { return name; }
}

class Group {
    String id;
    String name;
    List<User> members;
    List<Expense> expenses;

    public Group(String id, String name) {
        this.id = id; this.name = name;
        this.members = new ArrayList<>();
        this.expenses = new ArrayList<>();
    }
    public void addMember(User u) { members.add(u); }
    public void addExpense(Expense e) { expenses.add(e); }
}

class Expense {
    String id;
    String description;
    double amount;
    User paidBy;
    SplitType splitType;
    List<Split> splits;

    public Expense(String id, String desc, double amount, User paidBy, SplitType type, List<Split> splits) {
        this.id = id; this.description = desc; this.amount = amount;
        this.paidBy = paidBy; this.splitType = type; this.splits = splits;
    }
}

// ==========================================
// Service (Singleton)
// ==========================================

class SplitwiseService {
    private static SplitwiseService instance;
    private Map<String, User> users;
    private Map<String, Group> groups;
    
    // Balance Sheet: Map<User1, Map<User2, Double>>
    // User1 owes/gets User2 Amount. 
    // If > 0, User1 gets from User2. If < 0, User1 owes User2.
    private Map<String, Map<String, Double>> balanceSheet;

    private SplitwiseService() {
        users = new ConcurrentHashMap<>();
        groups = new ConcurrentHashMap<>();
        balanceSheet = new ConcurrentHashMap<>();
    }

    public static synchronized SplitwiseService getInstance() {
        if(instance == null) instance = new SplitwiseService();
        return instance;
    }

    public void addUser(User u) { 
        users.put(u.getId(), u); 
        balanceSheet.put(u.getId(), new HashMap<>());
    }
    public void addGroup(Group g) { groups.put(g.id, g); }

    public void addExpense(String desc, double amount, User paidBy, List<Split> splits, SplitType type) {
        // Validate Splits based on Type
        if(type == SplitType.EQUAL) {
            double splitAmount = amount / splits.size();
            for(Split s : splits) s.setAmount(splitAmount);
        } 
        else if (type == SplitType.PERCENT) {
            for(Split s : splits) {
                PercentSplit ps = (PercentSplit) s;
                double splitAmount = (amount * ps.getPercent()) / 100.0;
                s.setAmount(splitAmount);
            }
        }
        // For EXACT, amount is already set in Split
        
        // Update Balances
        for(Split split : splits) {
            User paidTo = split.getUser();
            double splitAmt = split.getAmount();
            
            // Should not add debt to self
            if(!paidBy.getId().equals(paidTo.getId())) {
                updateBalance(paidBy.getId(), paidTo.getId(), splitAmt);
                updateBalance(paidTo.getId(), paidBy.getId(), -splitAmt);
            }
        }
        
        System.out.println("Expense Added: " + desc);
    }
    
    // Settlement: User 'paidBy' pays 'paidTo' to clear debt
    public void settleBalance(User paidBy, User paidTo, double amount) {
        updateBalance(paidBy.getId(), paidTo.getId(), amount); // Payer is 'giving' money, so they get positive offset vs Payee
        updateBalance(paidTo.getId(), paidBy.getId(), -amount);
        System.out.println("Settled: " + paidBy.getName() + " paid " + amount + " to " + paidTo.getName());
    }
    
    private void updateBalance(String u1, String u2, double amount) {
        Map<String, Double> u1Balances = balanceSheet.get(u1);
        u1Balances.put(u2, u1Balances.getOrDefault(u2, 0.0) + amount);
    }

    public void showBalance(String userId) {
        Map<String, Double> balances = balanceSheet.get(userId);
        boolean isEmpty = true;
        for(Map.Entry<String, Double> entry : balances.entrySet()) {
            // Only show non-zero balances
            if(Math.abs(entry.getValue()) > 0.01) { 
                isEmpty = false;
                printBalance(userId, entry.getKey(), entry.getValue());
            }
        }
        if(isEmpty) System.out.println("No balances for " + users.get(userId).getName());
    }

    private void printBalance(String u1, String u2, double amount) {
        String user1Name = users.get(u1).getName();
        String user2Name = users.get(u2).getName();
        if(amount > 0) {
            System.out.println(user1Name + " gets " + String.format("%.2f", amount) + " from " + user2Name);
        } else if (amount < 0) {
            System.out.println(user1Name + " owes " + String.format("%.2f", Math.abs(amount)) + " to " + user2Name);
        }
    }
}

File: LLD/SplitwiseSystem.py
# fmt: off
# ==============================================================================
#  SPLITWISE SYSTEM — ASCII CLASS DIAGRAM
# ==============================================================================
#
#  ┌──────────────────────────────────────────────────────────────────────────┐
#  │                         SPLITWISE SYSTEM                                 │
#  └──────────────────────────────────────────────────────────────────────────┘
#
#  ┌─────────────────────────────────────────────┐
#  │              SplitwiseService               │  ← Facade
#  ├─────────────────────────────────────────────┤
#  │ + users: Dict[str, User]                    │
#  │ + expenses: List[Expense]                   │
#  │ + balances: Dict[str, Dict[str, Decimal]]   │
#  │ - _lock: Lock                               │
#  ├─────────────────────────────────────────────┤
#  │ + add_user()                                │
#  │ + add_expense(payer, total, split, members) │
#  │ + settle_balance(paid_by, paid_to, amount)  │
#  │ + show_balances()                           │
#  └─────────────────────────────────────────────┘
#                     │
#                     │ creates
#                     ▼
#  ┌─────────────────────────────────────────────┐
#  │                 Expense                     │
#  ├─────────────────────────────────────────────┤
#  │ + id: str                                   │
#  │ + description: str                          │
#  │ + amount: Decimal                           │
#  │ + paid_by: User                             │
#  │ + split: SplitStrategy                      │
#  │ + members: List[User]                       │
#  ├─────────────────────────────────────────────┤
#  │ + compute_shares() -> Dict[User, Decimal]   │
#  └─────────────────────────────────────────────┘
#                     │ 1
#                     ▼
#  ┌─────────────────────────────────────────────┐
#  │           SplitStrategy (ABC)               │  ← Strategy Pattern
#  ├─────────────────────────────────────────────┤
#  │ + compute(amount, members): Dict[User,Dec]  │
#  └──────────────┬──────────────────────────────┘
#                 │
#       ┌─────────┼──────────┐
#       ▼         ▼          ▼
#  ┌─────────┐ ┌────────┐ ┌──────────────────────┐
#  │ Equal   │ │ Exact  │ │    PercentSplit       │
#  │  Split  │ │ Split  │ ├──────────────────────┤
#  ├─────────┤ ├────────┤ │ percentages: Dict    │
#  │ amount/n│ │fixed   │ │ (must sum to 100%)   │
#  │ each    │ │amounts │ └──────────────────────┘
#  └─────────┘ └────────┘
#
#  ┌──────────────────────────────┐
#  │            User              │
#  ├──────────────────────────────┤
#  │ + id: str                    │
#  │ + name: str                  │
#  │ + email: str                 │
#  └──────────────────────────────┘
#
#  Balance Map Structure:
#  balances[user_a_id][user_b_id] = X
#  → means user_a is OWED X by user_b (user_b owes user_a)
#  → Positive = user_a gets money FROM user_b
#  → Negative = user_a OWES user_b
#
#  RELATIONSHIPS:
#  SplitwiseService ──*──> User           (registered users)
#  SplitwiseService ──*──> Expense        (recorded expenses)
#  Expense ──1──> SplitStrategy          (how to split the bill)
#  Expense ──1──> User (payer)           (who paid)
#  Expense ──*──> User (members)         (who shares the split)
#  EqualSplit ──▷── SplitStrategy        (implements: amount/n each)
#  ExactSplit ──▷── SplitStrategy        (implements: fixed amounts)
#  PercentSplit ──▷── SplitStrategy      (implements: % based)
#  Thread-safe: _lock guards balances dict and expense list
# ==============================================================================
# fmt: on
import threading
import uuid
from abc import ABC, abstractmethod
from decimal import Decimal, ROUND_HALF_UP
from enum import Enum
from typing import List, Dict, Optional

"""
==============================================================================================
SPLITWISE LOW LEVEL DESIGN (INTERVIEW OPTIMIZED)
==============================================================================================

Key Features:
1. User/Group Management: Create users, groups, and manage membership.
2. Expense Management: Supports Equal, Exact, and Percentage splits.
3. Balance Calculation: Tracks "Who owes Whom" across users.
4. Settlement: Logic to clear debts between users.
5. Concurrency: Thread-safe balance updates.
6. Precision: Decimal for financial accuracy.

Design Patterns:
1. Facade: SplitwiseService (Central Controller).
2. Strategy/Polymorphism: Split types (Equal, Exact, Percent).

Class Design Diagram:
---------------------
[SplitwiseService] "1" *-- "*" [User]
[SplitwiseService] "1" *-- "*" [Group]
[Group] "1" *-- "*" [Expense]
[Expense] "1" *-- "*" [Split]
[Expense] "1" *-- "1" [User] (PaidBy)
[Split] <|-- [EqualSplit]
[Split] <|-- [ExactSplit]
[Split] <|-- [PercentSplit]
[User] ..> [BalanceSheet] (Map<User, Decimal>)

Class Details:
---------------------
1. SplitwiseService (Facade)
   - Role: Main controller.
   - Methods: addExpense(), settleBalance(), showBalances().

2. Expense
   - Role: Represents a financial transaction shared between users.
   - Attributes: amount, paidBy, splits (List), splitType.

3. Split (Abstract)
   - Role: Represents a share of an expense for one user.
   - Subclasses: EqualSplit, ExactSplit, PercentSplit.

4. SplitType (Enum)
   - Types: EQUAL, EXACT, PERCENT.

5. User
   - Role: Participant.
   - Attributes: id, name, email.
"""

# ==========================================
# Enums & Models
# ==========================================

class SplitType(Enum):
    EQUAL = "EQUAL"
    EXACT = "EXACT"
    PERCENT = "PERCENT"

class User:
    def __init__(self, user_id: str, name: str, email: str = ""):
        self.id = user_id
        self.name = name
        self.email = email

    def __repr__(self):
        return f"User({self.name})"

# ==========================================
# Split Strategy Pattern
# ==========================================

class Split(ABC):
    def __init__(self, user: User, amount: Decimal = Decimal("0.00")):
        self.user = user
        self.amount = amount

class EqualSplit(Split):
    def __init__(self, user: User):
        super().__init__(user)

class ExactSplit(Split):
    def __init__(self, user: User, amount: Decimal):
        super().__init__(user, amount)

class PercentSplit(Split):
    def __init__(self, user: User, percent: float):
        super().__init__(user)
        self.percent = percent

# ==========================================
# Domain Models
# ==========================================

class Expense:
    """Represents a financial transaction shared between users."""
    def __init__(self, desc: str, amount: Decimal, paid_by: User, splits: List[Split], split_type: SplitType):
        self.id = str(uuid.uuid4())
        self.description = desc
        self.amount = amount
        self.paid_by = paid_by
        self.splits = splits
        self.split_type = split_type

class Group:
    def __init__(self, group_id: str, name: str):
        self.id = group_id
        self.name = name
        self.members: Dict[str, User] = {}
        self.expenses: List[Expense] = []

    def add_member(self, user: User):
        self.members[user.id] = user

    def add_expense(self, expense: Expense):
        self.expenses.append(expense)

# ==========================================
# Service Manager (Facade)
# ==========================================

class SplitwiseService:
    """Facade for managing users, groups, and expenses."""
    def __init__(self):
        self.users: Dict[str, User] = {}
        self.groups: Dict[str, Group] = {}
        # Balance Sheet: Map[User1_ID, Map[User2_ID, Decimal]]
        # Positive: User1 is owed by User2. Negative: User1 owes User2.
        self.balance_sheet: Dict[str, Dict[str, Decimal]] = {}
        self._lock = threading.Lock()
        print("INFO: SplitwiseService initialized.")

    def add_user(self, user: User):
        self.users[user.id] = user
        self.balance_sheet[user.id] = {}

    def add_group(self, group: Group):
        self.groups[group.id] = group

    def add_expense(self, desc: str, amount: Decimal, paid_by: User, splits: List[Split], split_type: SplitType):
        """Processes an expense and updates balances across all participants."""
        self._calculate_splits(amount, splits, split_type)

        with self._lock:
            for split in splits:
                paid_to = split.user
                if paid_by.id == paid_to.id:
                    continue
                # paidBy gets back split.amount from paid_to
                self._update_balance(paid_by.id, paid_to.id, split.amount)
                # paid_to owes split.amount to paidBy
                self._update_balance(paid_to.id, paid_by.id, -split.amount)

        print(f"INFO: Expense '{desc}' - {amount} paid by {paid_by.name}")

    def _calculate_splits(self, total: Decimal, splits: List[Split], split_type: SplitType):
        if split_type == SplitType.EQUAL:
            split_amt = (total / len(splits)).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
            for s in splits:
                s.amount = split_amt
            # Fix rounding remainder on first split
            total_calc = split_amt * len(splits)
            if total_calc != total:
                splits[0].amount += (total - total_calc)

        elif split_type == SplitType.PERCENT:
            for s in splits:
                if isinstance(s, PercentSplit):
                    s.amount = (total * Decimal(str(s.percent)) / Decimal("100")).quantize(Decimal("0.01"), rounding=ROUND_HALF_UP)
        # EXACT: s.amount is already provided

    def _update_balance(self, u1_id: str, u2_id: str, amount: Decimal):
        balances = self.balance_sheet[u1_id]
        balances[u2_id] = balances.get(u2_id, Decimal("0.00")) + amount

    def settle_balance(self, paid_by: User, paid_to: User, amount: Decimal):
        """Processes a settlement transaction between two users."""
        with self._lock:
            self._update_balance(paid_by.id, paid_to.id, amount)
            self._update_balance(paid_to.id, paid_by.id, -amount)
        print(f"INFO: Settled: {paid_by.name} paid {amount} to {paid_to.name}")

    def show_balances(self, user_id: str):
        """Prints all debts/receivables for a specific user."""
        balances = self.balance_sheet.get(user_id, {})
        user_name = self.users[user_id].name
        non_zero = [(self.users[oid].name, amt) for oid, amt in balances.items() if abs(amt) > Decimal("0.00")]
        if not non_zero:
            print(f"INFO: No pending balances for {user_name}.")
            return
        for other_name, amount in non_zero:
            if amount > 0:
                print(f"INFO: {user_name} gets back ${amount} from {other_name}")
            else:
                print(f"INFO: {user_name} owes ${abs(amount)} to {other_name}")

# ==========================================
# Main Execution / Demo
# ==========================================

if __name__ == "__main__":
    print("--- Starting Splitwise System Demo ---")

    service = SplitwiseService()

    alice = User("U1", "Alice")
    bob = User("U2", "Bob")
    charlie = User("U3", "Charlie")
    service.add_user(alice)
    service.add_user(bob)
    service.add_user(charlie)

    print("\n[Scenario 1] Alice pays 300 for Lunch (Equal split with Bob & Charlie)")
    splits1 = [EqualSplit(alice), EqualSplit(bob), EqualSplit(charlie)]
    service.add_expense("Lunch", Decimal("300"), alice, splits1, SplitType.EQUAL)
    service.show_balances("U2")  # Bob owes Alice 100

    print("\n[Scenario 2] Bob pays 100 for Cab (Alice owes 30, Bob keeps 70)")
    splits2 = [ExactSplit(alice, Decimal("30")), ExactSplit(bob, Decimal("70"))]
    service.add_expense("Cab", Decimal("100"), bob, splits2, SplitType.EXACT)
    service.show_balances("U1")  # Alice owes Bob 30

    print("\n[Scenario 3] Alice pays 100 to Bob to settle partially.")
    service.settle_balance(alice, bob, Decimal("100"))
    service.show_balances("U1")
    service.show_balances("U3")  # Charlie still owes Alice 100
